### ISSUE TITLE
The latest 3 commits that need to be incorporated somehow.

### ISSUE THREAD

**Issue by xlab** [cznic/cc#2]
Hi, please review commits b58b6706c0cb056df7341f5d3d8aa633bf469311, 0aefe3d846785a9312c42e89a86a55377cc5048f and 37d1391ad4ead6c249062da3baaa7c80169e93bf in the `cgogen` branch. These are only changes I made to keep things running, so finishing them off will mean a **full sync** with upstream master of the CC, that'd be a very happy moment for me.

So, the patch regarding `TypeSpecifier() *TypeSpecifier` allows me to walk enums and get their members, since there is no native support of enums in `Type` yet.

Then, there is no implemented `sizeOf / alignOf / structAlignOf` for enums, so I added a few methods to walk enum and get the top type (because it could upgrade from `uint32` up to `int64`). And use its values according to the model. This is very naive and may not work as per C specs, just for a consideration and as a temporary solution.

**Both commits above are not for merging!**

But the latest one is about a missing case in `MustConvert`, it gets triggered in cases like:
```c
#define ABC 1000.0f
```

That seems to be all.

**Comment by xlab** [comment]
Much thanks for separating CC out of C repo! 🍻

**Comment by cznic** [comment]
I'll push to the `cgogen` branch in few minutes. Discussion:

https://github.com/cznic/cc/commit/b58b6706c0cb056df7341f5d3d8aa633bf469311: I've made the access to enum constants a bit more higher level. Type `Type` is extended with a new method `EnumeratorList`, which, if type's `Kind == Enum`, returns a list of all of the type's `EnumConstant`s, if any.

https://github.com/cznic/cc/commit/0aefe3d846785a9312c42e89a86a55377cc5048f: I've made enum types have always the properties of C int. The reason is that the C99 specification mandates that the enum const expression value must be representable as C int.

https://github.com/cznic/cc/commit/37d1391ad4ead6c249062da3baaa7c80169e93bf: Small change.

In summary: Please review the upcoming commit and let me know, if the API addition, as well as the other changes, are ok for you.

Thanks for you help!

**Comment by xlab** [comment]
Regarding 7aba5d3a0c9b3a05073ed7e4f680af7d9f82b736: I'd like to have the original tokens of enum expression, sometimes it is possible to use them in Go code as-is.

An example: the fields from enumeration
```
0x04 << 1,
0x04 << 2,
0x04 << 3,
```
will look x100 better than their actual values resolved.

**Comment by cznic** [comment]
We *can* extend `EnumConstant` to have a `[]xc.Token` of its respective `ConstantExpression`. However, notice it's a bit slippery slope, because the operator precedence of C is not the same as of Go, ie. writing out `TokSrc()`s of the list may yield a different value than expected.

Moreover, it's not easy to reconstruct the token sequence of a `ConstantExpression`, because it has already became an AST, which must be walked and have a switch case for every `Expression.Case`. There are 50+ of those cases :-(

Your opinions?

**Comment by xlab** [comment]
That was exactly the case why I decided to finally sync our branches: I had my own simple code for getting values of macros, but then I got defines like `1000.0f` and even `~(0U)` so I decided to do all the refactorings and finally switch toy your macro map. However, I left intact the code that tries to convert a macro expression into a Go expression, because in case of PortAudio it causes no damage 100%.

My generator uses a config when you can explicitly say

```yml
TRANSLATOR: 
  ConstRules: 
    defines: expand
    enum: cgo
```

By default defines are evaluated (i.e. used with the value resolved by CC) so this is 100% safe. But in cases like PortAudio and many others (many, many others) it's better to keep the original expression because the sequence is correct in Go and it has some additional clarification (like cost cross-reference:  `1 + B` where B is also a const defined somewhere above). This is called `expand`.

**Comment by cznic** [comment]
So what shall we do? Walk the `ConstantExpression` AST and reconstruct the token sequence the preprocessor has produced? Any macro names will be already lost (expanded) anyway, but constant names will be still present.

I can do that, I just attempted to avoid the bit harder work, because I'm naturally lazy ;-)

**Comment by xlab** [comment]
All the hard work in this case should be done by the user of CC, like I did: https://github.com/xlab/cgogen/blob/master/translator/translator.go#L303

I only ask for a way to get the tokens from the original expression. I used to have it this way (for enums):

```
func (t *Translator) walkEnumerator(e *cc.Enumerator) {
	switch e.Case {
	case 0: // EnumerationConstant
	case 1: // EnumerationConstant '=' ConstantExpression
		decl.Value = e.ConstantExpression.Value
		decl.Expression = string(e.ConstantExpression.Expression.Token.S()) // ta-daa
	}
}
```

**Comment by cznic** [comment]
Okay, you gonna get the token list. I think I see a way how to do that w/o walking the AST ;-)

**Comment by xlab** [comment]
NB: I think methods like this would get exported eventually, but that's another story.

```c
func (m *Model) floatConst(lx *lexer, t xc.Token) (interface{}, Type) {
	s0 := t.S()
	s := s0
	switch s[len(s)-1] {
	case 'l', 'L', 'f', 'F':
		s = s[:len(s)-1]
	}
	f, err := strconv.ParseFloat(string(s), 64)
	if err != nil {
		lx.report.Err(t.Pos(), "invalid floating point constant")
		f = 0
	}
	switch s0[len(s0)-1] {
	case 'l', 'L':
		return m.value2(f, m.LongDoubleType)
	case 'f', 'F':
		return m.value2(f, m.FloatType)
	default:
		return m.value2(f, m.DoubleType)
	}
}
```

**Comment by cznic** [comment]
> NB: I think methods like this would get exported eventually, but that's another story.

Sure, probably with some small changes, like returning an error instead of reporting it to `lx.report`. So the error-returning variant would be exported and the "reporting" version would call the error-returning version.

Feel free to have a stab on it ;-)

**Comment by xlab** [comment]
```go
for _, en := range typ.EnumeratorList() {
 log.Println(en.ID) // always 0
}
```

**Comment by cznic** [comment]
> // always 0

Will fix.

**Comment by cznic** [comment]
Fixed ;-)

**Comment by xlab** [comment]
`e.ConstantExpression.toks` can be nil:

```
goroutine 1 [running]:
panic(0x375820, 0xc82000e100)
	/usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/cznic/cc.(*ctype).EnumeratorList(0xc8203a24e0, 0xc8204aaa20, 0x2, 0x2)
	/Users/xlab/Documents/dev/go/src/github.com/cznic/cc/etc.go:971 +0x1de
```

**Comment by xlab** [comment]
Wouldn't be this consistent with macros, if we replace `ID` field with `DefTok` actually?

**Comment by cznic** [comment]
DefTok: Yes, sounds reasonable to me.

PS: Away from a PC, will review your changes probably tomorrow.

**Comment by cznic** [comment]
And thanks for looking into this!

**Comment by cznic** [comment]
Reviewed https://github.com/cznic/cc/commit/afbc6893a44465839e6eb1cedff43dba256507cc, LGTM.

**Comment by cznic** [comment]
@xlab Max, this branch seems to be ready for merging at this moment. Please let me know what you think. Or please feel free to merge by yourself if you agree. Thank you.

**Comment by xlab** [comment]
I'm good! So far everything works, I'll recheck after your merge.

**Comment by cznic** [comment]
I'll change ID to DefTok in minutes and then merge the branch.

**Comment by cznic** [comment]
Merged. Please test and close if ok. Thank you.

**Comment by xlab** [comment]
@cznic 
Thanks man! Everything seems right, I've switched to the master.

**Comment by cznic** [comment]
> Everything seems right, I've switched to the master.

Congrats! 😉
